File: 1032 Sharing.py
Nodes = [[None, -1, 0] for _ in range(100000)]   # [key, next, flag]
S1, S2, N = map(int, input().split())

if __name__ == "__main__":
    for _ in range(N):
        add, key, next = input().split()
        Nodes[int(add)][0:2] = key, int(next)
        
    node = S1
    while node != -1:
        Nodes[node][2] = 1
        node = Nodes[node][1]
    
    node = S2
    while node != -1:
        if Nodes[node][2] == 1:
            print(f"{node:>05d}")
            exit(0)
        node = Nodes[node][1]
    
    print(-1)


'''
5 运行超时 0 -- 19188 KB
'''

File: 1032 Sharing.cpp
#include<cstdio>
using namespace std;

struct Chr{
    char c;
    int next;
    int flag = 0;
} a[100001];

int n, ah, bh;

int main(){
    scanf("%d %d %d", &ah, &bh, &n);
    int i;
    
    
    for (i=0; i<n; i++){
        int id;
        scanf("%d ", &id);
        scanf("%c %d", &a[id].c, &a[id].next);
    }
    
    int p = ah;
    while (p != -1){
        a[p].flag = 1;
        p = a[p].next;
    }
    
    p = bh;
    while (p != -1){
        if (a[p].flag == 1){
            printf("%05d\n", p);
            return 0;
        }
        p = a[p].next;
    }
    
    printf("-1\n");
    
    return 0;
}
